x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/CombineProjections.java

```diff
@@ -137,23 +141,33 @@ private static List<NamedExpression> combineProjections(List<? extends NamedExpr
     }
 
     private static List<Expression> combineUpperGroupingsAndLowerProjections(
-        List<? extends Attribute> upperGroupings,
+        List<? extends NamedExpression> upperGroupings,
         List<? extends NamedExpression> lowerProjections
     ) {
         // Collect the alias map for resolving the source (f1 = 1, f2 = f1, etc..)
-        AttributeMap<Attribute> aliases = new AttributeMap<>();
+        AttributeMap<Expression> aliases = new AttributeMap<>();
         for (NamedExpression ne : lowerProjections) {
-            // Projections are just aliases for attributes, so casting is safe.
-            aliases.put(ne.toAttribute(), (Attribute) Alias.unwrap(ne));
+            // record the alias
+            aliases.put(ne.toAttribute(), Alias.unwrap(ne));
         }
```

Comment:
nit: I do not think this change is necessary. The lower projections should still only be aliases that point to attributes - that's all a projection is allowed to contain (even though we don't make this invariant explicit).

Reply:
@ivancea Alex had a few comments on this file. Did y'all finish these? Or should these also go to a follow-up?

Reply:
Should be addressed in https://github.com/elastic/elasticsearch/pull/117882